coreldraw_cdr.ksy

```diff
@@ -2121,15 +2165,456 @@ types:
+                type: u4
+              - id: char_descriptions
+                type: char_description
+                repeat: expr
+                repeat-expr: num_chars
+              - id: num_bytes_in_text
+                type: u4
+              - id: text_data
+                size: num_bytes_in_text
+              - id: has_path_raw
+                type: u1
+              - size: num_chars * 24
+                if: has_path
+            instances:
+              has_path:
+                value: has_path_raw != 0
+          style_record:
+            seq:
+              - id: st_flag_1
+                type: u2
+              - id: st_flag_2
+                type: u2
+              - id: st_flag_3
+                type: u2
+              - id: url_properties
+                type: url_props
+                if: st_flag_2 == 0x3fff and (st_flag_3 & 0x11) == 0x11
+              - id: language
+                type: text_language
+                if: (st_flag_3 & 0x04) != 0
+              - id: style
+                type: style_string
+                if: st_flag_2 != 0 or (st_flag_3 & 0x04) != 0
+          skip:
+            seq:
+              - size: 16
+              - id: t_len
+                type: u4
+                # NOTE: libcdr checks for versions > 1600 instead, which seems like a mistake.
+              - size: '_root.version >= 1700 ? t_len : t_len * 2'
+      char_description:
+        seq:
+          - id: flags
+            type: u2
+          - id: style_override_idx_raw
+            type: u1
+          - size: 1 # usually 0x00, but can also be 0x20, 0x40 or 0x60
+          - size: 4
+            if: _root.version >= 1200
+        instances:
+          style_override_idx:
+            value: style_override_idx_raw >> 1
+      url_props:
+        seq:
+          - id: url_id_len
+            type: u4
+          - id: url_id_old
+            size: url_id_len * 2
+            type: str
+            encoding: UTF-16LE
+            if: _root.version < 1700
+          - id: url_id_new
+            size: url_id_len
+            type: str
+            encoding: ASCII
+            if: _root.version >= 1700
+        instances:
+          url_id_raw:
+            value: '_root.version < 1700 ? url_id_old : url_id_new'
+          url_id:
+            value: url_id_raw.to_i
+      text_language:
+        seq:
+          - id: value_old
+            size: 4
+            type: strz
+            encoding: ASCII
+            if: _root.version < 1300
+
+          - id: value_new_len
+            type: u4
+            if: _root.version >= 1300
+          - id: value_new
+            size: value_new_len.as<u4> * 2
+            type: str
+            encoding: UTF-16LE
+            if: _root.version >= 1300
+            doc-ref: https://www.ibm.com/docs/en/cics-ts/5.5?topic=development-national-language-codes-application
+        instances:
+          value:
+            value: '_root.version >= 1300 ? value_new : value_old'
```

Comment:
It probably doesn't make much sense to provide this instance, because the meaning of `value_old` and `value_new` is apparently slightly different. `value_old` seems to use 2-character codes, whereas `value_new` uses 3-character codes (as you indicate in its `doc-ref`) - this is an extract from sample file dumps:

```console
$ grep -rhF '"language"' -A2 | grep -ho '"value_old": "\?[^",]*"\?' | sort | uniq -c | sort -r
    366 "value_old": "US"
    148 "value_old": "RU"
      1 "value_old": "GR"
      1 "value_old": "EN"
      1 "value_old": "CE"

$ grep -rhF '"language"' -A2 | grep -ho '"value_new": "\?[^",]*"\?' | sort | uniq -c | sort -r
   1573 "value_new": "ENU"
     58 "value_new": "RUS"
     44 "value_new": "UKR"
     39 "value_new": "ESH"
     24 "value_new": "CSY"
     15 "value_new": "ENG"
     11 "value_new": "ENA"
      6 "value_new": "IND"
      5 "value_new": "ESM"
      4 "value_new": "PTB"
      3 "value_new": "ENC"
      3 "value_new": "DEU"
      2 "value_new": "LTH"
      2 "value_new": "ENS"
      1 "value_new": "HYE"
      1 "value_new": "DEA"
```

So perhaps it's better to keep `value_old` and `value_new` separate, because you'll need to handle them differently anyway. And `value_old` should also get a `doc-ref` or something with the list of the 2-character codes used, like `value_new` has.

Reply:
I actually don’t think it’s inconceivable that you’d want to handle them together, since AFAICT they’re disjoint sets that serve the same purpose (locale to use when decoding and/or presenting the text*). Still, it wouldn’t be hard for a user to combine the fields in their programming language later, so it’s not a big deal either way. Will add a `doc` or `doc-ref`.

*Now that I say that, I’m thinking I should rename `language` to the more general `locale`, `value_old` to `country` and `value_new` to `national_language`.

Reply:
Done.